### PR TITLE
Add Zstd/LZ4 support for region files (1.15 edition)

### PR DESCRIPTION
**Pull request by egg82**
Zstd using zstd-jni: https://github.com/luben/zstd-jni
LZ4 using lz4-java: https://github.com/lz4/lz4-java

Added new command-line option, --forceWrite, which will forcibly write the chunk/flush to disk regardless of whether or not the chunk has actually changed in a --forceUpgrade.
New compression only happens when a chunk is saved, so running a --forceUpgrade AND --forceWrite may be desirable.

New config option for the compression method to use.
There's a new option in PaperConfig (compression level) for Zstd which defaults to 6.
There's a new option in PaperConfig (compression level) for Zlib which defaults to 5.

Since the patch hooks Mojang's versioning system for new compression methods, it
should be internally consistent, easily-updatable, and reversible with a --forceUpgrade AND a --forceWrite.

This will likely conflict with plugins and programs expecting region files to be compressed using the current standard.

- [X] Implementation
- [X] Dictionary

### CODE REVIEW

**Comment by egg82** [comment]
Dictionary coming soon, as soon as I train one. This is the rest of the implementation completed.

Due to some updates in the Zstd library, this version should be even faster than the previous implementation!

**Comment by egg82** [comment]
Dictionary is done and I made some suggested corrections.

**Comment by egg82** [comment]
**Test results**
Ratio = "how many times can the compressed data fit into its own uncompressed data?" - so, higher is better.
Speed is in us, or microseconds. eg. 2000us = 2ms. Lower is better.
Dict (or dictionary) is with a dictionary trained on the uncompressed data.

Byte array decompression *can* be made more efficient by including the uncompressed size in the region/chunk headers, but using both byte arrays and adding the new header would require modification that is beyond the scope of this patch. Perhaps a future patch.

Zlib stream compression levels:
https://docs.google.com/spreadsheets/d/1eKatLMxA560RerGIRyYVWC8n4bkkRe4Uj0sBFg2wtOY/edit?usp=sharing
Zlib stream w/ dict compression levels:
https://docs.google.com/spreadsheets/d/1PVq4vWJWLyj-crCiB1nUdC13zc4HcCsPKotv5dLOU9c/edit?usp=sharing

Zstd stream compression levels:
https://docs.google.com/spreadsheets/d/1k0YL8nQEPYtV6pY1U_uRc2jfIz4WDObpmrM0KLWPvaM/edit?usp=sharing
Zstd stream w/ dict compression levels:
https://docs.google.com/spreadsheets/d/1P_75_VPDTjopefps26fgNnfspA-Zg7DEmkUVb-lQjz0/edit?usp=sharing

LZ4:
https://docs.google.com/spreadsheets/d/1iNcqo07ySM-r2-1DTBOmd5vl809u97AmacdxSOMiPCQ/edit?usp=sharing

Since 1.15 relies on streams, I decided not to fight it and let the ~15us penalty for streams over byte arrays go, as converting to a byte array would require more code, and thus be less maintainable down the line.

tl;dr from Zlib (current) to Zlib (new, default level) would mean:
1. An average compression ratio from ~9.001 to ~8.875
2. An average compression speed from ~651.813us to ~375.993us
3. An average decompression speed from ~89.391us to ~91.567us

from Zlib (new, default level) to Zstd (new, default level, with the included dictionary) would mean:
1. An average compression ratio from ~8.875 to ~21.299
2. An average compression speed from ~375.993us to ~346.219us
3. An average decompression speed from ~91.567us to ~40.155us

**Comment by Black-Hole** [comment]
How about a versioned dictionary? With future Minecraft versions it might be useful to update the dictionary.

File: Spigot-Server-Patches/0420-Add-Zstd-LZ4-support-for-region-files-1.15-edition.patch [hunk 1]
```diff
@@ -178,9 +183,47 @@ index 3382d678..a891b56f 100644
...
 +
-+    private static final com.github.luben.zstd.ZstdDictCompress zstdDictCompressor;
-+    private static final com.github.luben.zstd.ZstdDictDecompress zstdDictDecompressor;
-+    public static final RegionFileCompression ZSTD;
++    private static final short CURRENT_ZSTD_VERSION = 0;
++    private static final LoadingCache<Short, byte[]> zstdDicts = CacheBuilder.newBuilder().build(
++        new CacheLoader<Short, byte[]>() {
++            @Override
++            public byte[] load(Short key) throws Exception {
++                byte[] retVal = toBytes(RegionFileCompression.class.getClassLoader().getResourceAsStream("paper.zstd.dict.0." + key));
```
Comment (Black-Hole) [inline Spigot-Server-Patches/0420-Add-Zstd-LZ4-support-for-region-files-1.15-edition.patch]: ".0" is not right here.

Comment (egg82) [reply Spigot-Server-Patches/0420-Add-Zstd-LZ4-support-for-region-files-1.15-edition.patch]: Whoops, fixed.

**Review by Spottedleaf — APPROVED** [review]

**Comment by Spottedleaf** [comment]
gg

File: Spigot-Server-Patches/0420-Add-Zstd-LZ4-support-for-region-files-1.15-edition.patch [hunk 2]
```diff
@@ -0,0 +1,1059 @@
...
++        } else {
++            log("Using " + (regionCompressionMethod == RegionCompressionType.NONE ? "no" : regionCompressionMethod.getName()) + " compression for region files.");
++            regionCompressionZstdLevel = getInt("settings.region-compression.zstd-level", 6);
++            if (regionCompressionZstdLevel < Zstd.minCompressionLevel() || regionCompressionZstdLevel > Zstd.maxCompressionLevel()) {
++                fatal("Valid values for settings.region-compression.zstd-level are from " + Zstd.minCompressionLevel() + " to " + Zstd.maxCompressionLevel());
++            }
++            if (regionCompressionMethod == RegionCompressionType.ZSTD) {
++                log("Using level " + regionCompressionZstdLevel + " for Zstd compression." + (regionCompressionZstdLevel >= 14 ? " (Make my PC hurt)" : ""));
++            }
++            regionCompressionZlibLevel = getInt("settings.region-compression.zlib-level", 5);
```
Comment (electronicboy) [inline Spigot-Server-Patches/0420-Add-Zstd-LZ4-support-for-region-files-1.15-edition.patch:94]: should default from the field, same for the one above too

**Review by electronicboy — COMMENTED** [review]
I just wanna clarify here,what's the behavior of the zstd dicts if we bump? Do we have to keep the existing dict for decompression chunks or is the dict saved/whatever with the chunk data?

**Comment by Spottedleaf** [comment]
existing dict needs to be kept, only the dict version is stored on chunk data

**Comment by y6x2digc** [comment]
Any news about this?

**Comment by egg82** [comment]
Closed in favor of #5029
